File: ProjectPO_Java/src/PO_java/LifeSymulator/Animals/Antelope.java
package PO_java.LifeSymulator.Animals;

import PO_java.LifeSymulator.Animal;
import PO_java.LifeSymulator.Organism;
import PO_java.LifeSymulator.Tools.Cords;
import PO_java.LifeSymulator.World;

import java.util.Random;

public class Antelope extends Animal {
    private final int CHANCES = 2;
    private final int LUCKY = 0;
    public Antelope(World world, Cords cords)
    {
        strength = 4;
        initiative = 4;
        age = 0;
        sign = 'A';
        name = "Antelope";
        toKill = false;
        isDead = false;
        this.cords = cords;
        this.world = world;
    }

    @Override
    public void action(){
        int moveRange = 2; // zmiana zasiegu ruchu na 2
        Cords currentCords = this.cords;
        Cords newCords;
        newCords = new Cords(-1, -1);
        while (!checkMove(newCords))
            newCords.setCords(randomNewCords(currentCords, moveRange));
        if (checkMove(newCords))
        {
            if (world.isSomethingThere(newCords))
            {
                Organism occupyingOrganism = world.whatIsThere(newCords);
                collision(occupyingOrganism);
            }
            else
                move(this, newCords, currentCords);
        }
    }

    @Override
    public void defense(Organism attacker){
        Random rand = new Random();
        int chance = rand.nextInt(CHANCES);
        if(chance == LUCKY && checkFreeFieldsNear(this.cords)){
            Cords newCords = randomFreeFieldsNear(this.cords);
            move(this, newCords, this.cords);
            System.out.println("Antelope escaped from " + attacker.getName());
        }
        else{
            world.addOrganismToKill((this));
            move(attacker, this.cords, attacker.getCords());
            System.out.println(attacker.getName() + " ate Antelope");
        }
    }
}

File: ProjectCppPO/Antelope.cpp
#include "Antelope.h"
#define LUCKY 0

Antelope::Antelope(World* world, Cords cords)
{
	features.strength = 4;
	features.initiative = 4;
	features.age = 0;
	features.look = 'A';
	features.name = "Antelope";
	features.toKill = false;
	location = cords;
	this->world = world;
}

//Metoda nadpisuj¹ca metodê z klasy rodzica ¿eby antylopa mog³a sie ruszac o 2 pola
void Antelope::action()
{
	int moveRange = 2;
	Cords currentCords = this->getLocation();
	Cords newCords(-1, -1);
	while (!checkMove(newCords))
		newCords.setCords(randomNewCords(currentCords, moveRange));
	if (checkMove(newCords))
	{
		if (world->isSomethingThere(newCords))
		{
			Organism* occupyingOrganism = world->whatIsThere(newCords);
			collision(occupyingOrganism);
		}
		else
			move(this, newCords, currentCords);
	}
}

//Metoda obs³uguj¹ca mo¿liwoœc odskoczenia antylopy gdy ktoœ chce j¹ zaatakowaæ
void Antelope::defense(Organism* attacker)
{
	int chance = rand() % 2;
	if (chance == LUCKY && checkFreeFieldsNear(location))
	{
		Cords newCords = randomFreeFieldsNear(location);
		move(this, newCords, location);
		//dodanie raportu
		string report = features.name + " escaped from " + attacker->getFeatures().name;
		world->addReport(report);
	}
	else
	{
		world->addOrganismToKill(this);
		move(attacker, location, attacker->getLocation());
		//dodanie raportu
		string report = attacker->getFeatures().name + " ate " + features.name;
		world->addReport(report);
	}
}
